### ISSUE TITLE
PayPal Standard - Order confirmation page doesn't show Download section on a WordPress.com hosted site

### ISSUE THREAD

**Issue by thuautp** [woocommerce/woocommerce#30154]
**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
#4089467-zen

When purchasing a Downloadable product with PayPal Standard, the Order Confirmation/ Thank you page doesn't show the Download section anymore unless refreshing the page. I tested this on a Pressable site and couldn't replicate, however, I could on a WordPress.com eCom plan site.

I also tested with Stripe on the eCom site but also could not replicate, so this seems to be happening for PayPal Standard.

**Expected behavior**
Download section should show on the Order Confirmation page.

**Actual behavior**
Download section is missing on the Order Confirmation page, refreshing the page brought the section back up.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Create a Downloadable product on a WP.com eCom plan site
2. Set up PayPal Standard
3. Add product to cart and check out with PayPal Standard
4. Complete the checkout and return to Order Confirmation page, see that the Download section is missing
5. Refresh the page, the section should appear.

**Screenshots**

Before refreshing the page - WP.com eCom site:

![https://d.pr/MJQF9Y](https://d.pr/MJQF9Y+)
Full Image [here](https://d.pr/MJQF9Y+) 

After refreshing the page - WP.com eCom site:

![https://d.pr/YLUv0v](https://d.pr/YLUv0v+)
Full Image [here](https://d.pr/YLUv0v+) 

Test on Pressable site - can't replicate:

![https://d.pr/eZYzVi](https://d.pr/eZYzVi+)
Full Image [here](https://d.pr/eZYzVi+) 

**Isolating the problem (mark completed items with an [x]):**
- [x ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x ] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
We use the [WooCommerce System Status Report](https://docs.woocommerce.com/document/understanding-the-woocommerce-system-status-report/) to help us evaluate the issue. 
Without this report we won't be able to fully evaluate this issue.
<details>
```
 `
### WordPress Environment ###

WordPress address (URL): https://moggiesnmore.com
Site address (URL): https://moggiesnmore.com
WC Version: 5.4.1
REST API Version: ✔ 5.4.1
WC Blocks Version: ✔ 5.1.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 2.3.1
Log Directory Writable: ✔
WP Version: 5.7.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 7.4.20
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 7.77.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.4.19-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.4.1
WC Database Prefix: wp_
Total Database Size: 10.83MB
Database Data Size: 8.24MB
Database Index Size: 2.59MB
wp_woocommerce_sessions: Data: 0.27MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_ce4wp_abandoned_checkout: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.22MB + Index: 0.14MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_custom_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_dynamic_segment_filters: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_feature_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_forms: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_mapping_to_external_entities: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_newsletter_links: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_newsletter_option: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_option_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_segment: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_newsletter_templates: Data: 2.52MB + Index: 0.00MB + Engine InnoDB
wp_mailpoet_scheduled_tasks: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_scheduled_task_subscribers: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_segments: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_sending_queues: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_settings: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_clicks: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_statistics_newsletters: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_statistics_opens: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_mailpoet_statistics_unsubscribes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_mailpoet_statistics_woocommerce_purchases: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_mailpoet_stats_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_subscribers: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_mailpoet_subscriber_custom_field: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_ips: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailpoet_subscriber_segment: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_mailpoet_user_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.11MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 1.20MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.39MB + Index: 0.06MB + Engine InnoDB
wp_taxjar_record_queue: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 47
customize_changeset: 17
jetpack_migration: 2
jp_img_sitemap: 1
jp_sitemap: 1
jp_sitemap_master: 1
mailpoet_page: 1
nav_menu_item: 5
page: 7
post: 2
product: 2
revision: 47
shop_order: 12
wp_template_part: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (21) ###

Akismet Anti-Spam: by Automattic – 4.1.9
AMP: by AMP Project Contributors – 2.1.2
CoBlocks: by GoDaddy – 2.12.1
Creative Mail by Constant Contact: by Constant Contact – 1.4.0
Crowdsignal Forms: by Automattic – 1.5.3
Facebook for WooCommerce: by Facebook – 2.6.0
WordPress.com Editing Toolkit: by Automattic – 3.9166
Google Listings and Ads: by WooCommerce – 1.0.0
Gutenberg: by Gutenberg Team – 10.8.2
Jetpack: by Automattic – 9.8.1
Layout Grid: by Automattic – 1.6
Mailchimp for WooCommerce: by Mailchimp – 2.5.1
MailPoet 3 (New): by MailPoet – 3.64.1
Page Optimize: by Automattic – 0.5.1
Storefront Powerpack: by WooCommerce – 2.0.0
TaxJar - Sales Tax Automation for WooCommerce: by TaxJar – 3.2.10
WooCommerce Payments: by Automattic – 2.5.0
WooCommerce PayPal Payments: by WooCommerce – 1.3.2
WooCommerce Product Add-ons: by WooCommerce – 4.1.0
WooCommerce Shipping & Tax: by WooCommerce – 1.25.14
WooCommerce: by Automattic – 5.4.1

### Inactive Plugins (13) ###

Classic Editor: by WordPress Contributors – 1.6
Klarna Checkout for WooCommerce: by Krokedil – 2.5.7
Klarna Payments for WooCommerce: by krokedil
klarna
automattic – 2.5.1

WooCommerce Australia Post Shipping: by WooCommerce – 2.4.28
WooCommerce Canada Post Shipping: by WooCommerce – 2.5.19
WooCommerce eWAY Payment Gateway: by WooCommerce – 3.2.2
WooCommerce PayFast Gateway: by WooCommerce – 1.4.19
WooCommerce PayPal Checkout Gateway: by WooCommerce – 2.1.1
WooCommerce Royal Mail: by WooCommerce – 2.5.34
WooCommerce Square: by WooCommerce – 2.5.1
WooCommerce Stripe Gateway: by WooCommerce – 5.2.3
WooCommerce UPS Shipping: by WooCommerce – 3.2.32
WooCommerce USPS Shipping: by WooCommerce – 4.4.53

### Dropin Plugins (2) ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Must Use Plugins (1) ###

WP.com Site Helper: by  –

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: simple (simple)
Taxonomies: Product Visibility: 
Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #18 - /shop/
Cart: #19 - /cart/
Checkout: #20 - /checkout/
My account: #21 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Ashe
Version: 1.9.9.5.8 (update to version 2.0 is available)
Author URL: https://wp-royal.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 60
Oldest: 2021-06-22 16:55:42 +0100
Newest: 2021-06-23 17:05:28 +0100


### Status report information ###

Generated at: 2021-06-23 17:06:31 +01:00
`
```
</details>

**Comment by simplysaru** [comment]
@thuautp .I am able to replicate this issue on WordPress.com Business plan as well.  That said since this issue happens only with WordPress.com and Paypal Standard, I would suggest filing this bug report on the Paypal repository as this isn't a core issue, because this works as expected on Stripe and other payment providers.

**Comment by AashikP** [comment]
Hello @thuautp 

Thank you for reporting this bug.

I'm wondering if this is something that needs to be resolved on the WooCommerce side or WordPress.com.I'm adding the needs developer feedback here as this bug can be replicated consistently, but only on WordPress.com:

![https://d.pr/i/FLYZuF+](https://d.pr/i/FLYZuF+)

**Comment by Konamiman** [comment]
Hi @thuautp. Just a guess, but this seems to be caused by the fact that downloads are not allowed until order status is processing or completed, so at the time the "order received" screen shows maybe PayPal IPN hasn’t triggered the process yet; it's then handled later via webhook and thus on refresh the order status has changed and downloads are available.

To verify this, could you please try to replicate again but this time look at the order status at the time the "order received" screen shows?

**Comment by thuautp** [comment]
Hi @Konamiman , I've just tested this again but strangely can't replicate the issue anymore. I tested on the same site I used to test the issue originally ([link to product](https://testat2029.wpcomstaging.com/product/digital-download/)). I've just done testing twice and both times got the download links on the **Order received** screen, order status at the time is **Completed**.

![https://d.pr/i/gS8nrO](https://d.pr/i/gS8nrO+)
Full Image [here](https://d.pr/i/gS8nrO+) 

![https://d.pr/i/udBmAE](https://d.pr/i/udBmAE+)
Full Image [here](https://d.pr/i/udBmAE+) 

Can you still replicate the issue on your end?

**Comment by thuautp** [comment]
I've just tested for the third time, this time - no download link, order status is Completed. It looks like the issue seems intermittent?

![https://d.pr/i/HJSqFw](https://d.pr/i/HJSqFw+)
Full Image [here](https://d.pr/i/HJSqFw+) 

![https://d.pr/i/9dGVM4](https://d.pr/i/9dGVM4+)
Full Image [here](https://d.pr/i/9dGVM4+)

**Comment by melek** [comment]
Encountered in 4249657-zen.

I wonder if this might have to do with misconfigured IPN, as this user has a `example.com` receiver email: 
![image](https://user-images.githubusercontent.com/220149/131023714-56d2ab3d-76be-4b7d-9529-2c831f4ee0da.png)

Not sure how the Order details screen would be triggered before the downloadable product would be added to the account due to IPN failing, but I will troubleshoot IPN with the customer to see if that alleviates the issue.

Update: The user has active subscriptions, so I'm not recommending changing IPN settings - and their unusual email address does match their PayPal email. 

Though it isn't usually a good idea, I'll see if enabling PDT alongside IPN changes the issue at all.

**Comment by rashedripon** [comment]
> this seems to be caused by the fact that downloads are not allowed until order status is processing or completed, so at the time the "order received" screen shows maybe PayPal IPN hasn’t triggered the process yet; it's then handled later via webhook and thus on refresh the order status has changed and downloads are available.

I was able to reproduce the problem in my pressable site and the explanation by @Konamiman seems to be the correct one. 

I checked the order status as soon as I was on the order-received page (no "Downloads" available), the order status was "Pending Payment". I reloaded the order status page after a few seconds and it was "Completed". Then I refreshed the order-received page and I can see the "Downloads" section.

It seems the webhook is taking time to change the order status in WooCommerce. Can we make it faster or is there a solution that we can suggest for now?

I have tested with [PayPal Payments](https://woocommerce.com/products/woocommerce-paypal-payments/) and there was no delays to update the order status. Also, I got the "Downloads" section immediately, refreshing the page was not required.

**Comment by shohanhossainnab** [comment]
Hi! Just wanted to check if there's any update on this?

**Comment by barryhughes** [comment]
> I've just tested for the third time, this time - no download link, order status is Completed. It looks like the issue seems intermittent?

Without further info I can't say for sure, but strongly suspect that this is a timing issue, with the order received page rendering just slightly ahead of the order being updated to 'completed' (and that latter action is perhaps delayed because of latency with IPN: not something we can control).

> It seems the webhook is taking time to change the order status in WooCommerce. Can we make it faster or is there a solution that we can suggest for now?

By webhook I think you mean the IPN message (but let me know if I'm misunderstanding). Unfortunately, there is no way for us to expedite that: it's entirely controlled by PayPal.

Of course, you can also enable PDT which means that as soon as the customer is returned to the merchant website, updated information will be available almost right away.

---

There doesn't seem to be enough information here to suggest there is an actual bug. What I'd strongly recommend is (unless you are already using it) is enabling PDT (this can be done in addition to IPN, it doesn't replace it).

To that end, I'm going to go ahead and close this.

**Comment by eevolution** [comment]
I have the same issue today with a new site when testing using Stripe (in Test Mode) on AWS Lightsail. Can a setting be added so we can specify which process or hook needs to be completed before showing the order received page? Alternatively, the downloads section continually looks for the file(s) using ajax and refreshes informing the customer when it's ready?

**Comment by barryhughes** [comment]
Hi! This issue has been closed, so we may not always see any updates you post.

> Can a setting be added so we can specify which process or hook needs to be completed before showing the order received page? 

I'm not completely sure what you are proposing (or what should happen if a certain hook isn't triggered or specific stage in the payment process is not reached), but do feel free to create a new issue with details and we can discuss/evaluate from there :+1:

> Alternatively, the downloads section continually looks for the file(s) using ajax and refreshes informing the customer when it's ready?

This feels like customization territory, and it is unlikely we would add it at this time.
